cascada/primitives/rectangle.py

```diff
@@ -0,0 +1,219 @@
+"""RECTANGLE cipher."""
+from cascada.bitvector.core import Constant
+from cascada.bitvector.operation import RotateLeft
+
+from cascada.bitvector.ssa import RoundBasedFunction
+from cascada.primitives.blockcipher import Encryption, Cipher
+
+
+def sub_column(x):
+    t0 = x[2]
+    x[2] ^= x[1]
+    x[1] = ~x[1]
+    t1 = x[0]
+    x[0] &= x[1]
+    x[1] |= x[3]
+    x[3] ^= t0
+    x[0] ^= x[3]
+    x[1] ^= t1
+    x[3] &= x[1]
+    x[3] ^= x[2]
+    x[2] |= x[0]
+    x[2] ^= x[1]
+    x[1] ^= t0
+
+
+def shift_row(x):
+    x[1] = RotateLeft(x[1], 1)
+    x[2] = RotateLeft(x[2], 12)
+    x[3] = RotateLeft(x[3], 13)
+
+
+def add_round_key(x, k, i):
+    x[0] ^= k[i]
+    x[1] ^= k[i + 1]
+    x[2] ^= k[i + 2]
+    x[3] ^= k[i + 3]
+
+
+RC = (0x01, 0x02, 0x04, 0x09, 0x12, 0x05, 0x0B, 0x16,
+      0x0C, 0x19, 0x13, 0x07, 0x0F, 0x1F, 0x1E, 0x1C,
+      0x18, 0x11, 0x03, 0x06, 0x0D, 0x1B, 0x17, 0x0E,
+      0x1D)
+
+
+class RECTANGLEEncryption(Encryption, RoundBasedFunction):
+    """RECTANGLE encryption function."""
+
+    num_rounds = 25
+    input_widths = [16, 16, 16, 16]
+    output_widths = [16, 16, 16, 16]
+    round_keys = None
+
+    @classmethod
+    def set_num_rounds(cls, new_num_rounds):
+        cls.num_rounds = new_num_rounds
+
+    @classmethod
+    def eval(cls, *x):
+        x = list(x)
+
+        for i in range(cls.num_rounds):
+            add_round_key(x, cls.round_keys, i * 4)
+            sub_column(x)
+            shift_row(x)
+            cls.add_round_outputs(x)
+
+        add_round_key(x, cls.round_keys, cls.num_rounds * 4)
+
+        return x
+
+
+class RECTANGLE80KeySchedule(RoundBasedFunction):
+    """RECTANGLE-80 key schedule function."""
+
+    num_rounds = 25
+    input_widths = [16, 16, 16, 16, 16]
+    output_widths = [16 for _ in range(4 * num_rounds + 4)]
+
+    @classmethod
+    def set_num_rounds(cls, new_num_rounds):
+        assert new_num_rounds <= len(RC)
+        cls.num_rounds = new_num_rounds
+        cls.output_widths = [16 for _ in range(4 * new_num_rounds + 4)]
+
+    @classmethod
+    def eval(cls, *master_key):
+        x = list(master_key)
+        round_keys = x[0:4]
+
+        for i in range(cls.num_rounds):
+            t = x[0:4]
+            sub_column(t)
+            x[0] = (x[0] & 0xFFF0) | (t[0] & 0x000F)
+            x[1] = (x[1] & 0xFFF0) | (t[1] & 0x000F)
+            x[2] = (x[2] & 0xFFF0) | (t[2] & 0x000F)
+            x[3] = (x[3] & 0xFFF0) | (t[3] & 0x000F)
```

Comment:
Can you replace these OR and AND operations with Extract and Concat operations? The advantage of using Extract and Concat is that they propagate properties with higher probabilities.

Reply:
Thanks, I did not know there are Extract and Concat operations.